### ISSUE TITLE
[SB2] terrible wifi performance when bluetooth is active

### ISSUE THREAD

**Issue by nqzero** [linux-surface/linux-surface#711]
i have the SB2 15" with ubuntu 21.04 and the custom surface kernel 5.15.6-surface
(i've tried several other combos as well)

with bluetooth disabled, the wifi works reasonably
with bluetooth enabled, the performance deteriorates
 - have seen rates of 24kbps
 - disconnects
 
this has been mentioned in many now-closed issues
however, the feature matrix still lists both bluetooth and wifi as supported

at the least, the matrix should be updated to reflect that the two cannot be used together
(i purchased the sb2 primarily to use the tablet without the base via a bluetooth keyboard and mouse in linux)
ideally, this (or a similar) bug report would also be left open

**Comment by jonas2515** [comment]
Well, for me those issues are not appearing anymore as long as https://github.com/linux-surface/kernel/pull/88 is applied (which it is in the linux-surface kernel since quite some time).

I assume you use 2.4Ghz? Have you paired a BT LE device like the surface pen (that would cause the firmware issues https://github.com/linux-surface/kernel/pull/88 works around)?

Also please double-check that the BT chip is not accidentally scanning all the time, there's a few bugs in gnome-control-center that might cause it to never stop scanning. Can check by looking for "Discovering: no" in the output of `bluetoothctl show`. If BT is discovering, very bad wifi performance is to be expected, because no time at all will be available for wifi traffic.

**Comment by Majiir** [comment]
I'm having this issue with a SP3 running NixOS 22.05 with `nixos-hardware/microsoft/surface-pro/3`. I have the issue with the 5.16.11 kernel referenced in nixos-hardware, and also the latest 5.16.17 from linux-surface/kernel. With bluetooth enabled, I only get 2-4 Mbps down and 18-20 Mbps up; but with bluetooth disabled, I get 100+ down and 50+ up. This is all on a 5GHz network. `bluetoothctl` outputs `Discovering: no`.

**Comment by NP-chaonay** [comment]
@Majiir 

- is Bluetooth is connected?
- have you try 2.4ghz?
- many laptoo may have this problem when Bluetooth is turned on, it interfere wifi, and wifi connection is less stable and slowed down
- have you tested on windows and also another linux?
- bluetoothctl output is not related to problem maybe and if it does then it not worse the problem, discovering no mean no other device will find this computer in Bluetooth connection pairing.

**Comment by Majiir** [comment]
> is Bluetooth is connected?

No, I experience the issue even when every paired device shows `Connected: no` in `bluetoothctl info <device>`.

But what's very interesting is that once I connected a device (my mouse, an MX Master) speeds improved again! Here are the results:

- Bluetooth off: 93 Mbps down / 41 Mbps up
- Bluetooth on, no devices connected: 3.1 Mbps down / 22.3 Mbps up
- Bluetooth on, mouse connected: 93 Mbps down / 39 Mbps up

All tests were performed using `speedtest.net` against the same server, on wifi which is connected to a router that handles 900+ Mbps up and down, in an area with very little RF interference. The above results are with both 5GHz and 2.4GHz enabled, and I get very similar results with 5GHz-only and 2.4GHz-only.

> have you tested on windows and also another linux?

I have not tested recently on Windows, but I did run Windows on this machine for years (up until ~6 months ago) and never had an issue with wifi speeds. It's a very impactful slowdown, so I think it's safe to say the issue doesn't happen on Windows. And my wifi setup was the same then as now.

I tried on the latest Fedora Workstation live USB, and I was able to replicate the issue there as well. (Note that I didn't install the surface-linux kernel; not sure if Fedora ships with it, but I doubt it?) It had the same issue where speeds were slow when bluetooth is enabled and nothing is connected, but speeds were fast when bluetooth was disabled _or_ a device was connected.

> if wifi is usable just less speed (and not much bery low speed which imo your less speed is not much that low) idk if it normal speed when you using it with bluetooth.

Dropping from 100+ Mbps to 3Mbps makes a huge difference for user experience. Latency also increases dramatically (from 50ms to 400ms).

**Comment by NP-chaonay** [comment]
> Note that I didn't install the surface-linux kernel; not sure if Fedora ships with it, but I doubt it?

No distro ship surface-kernel, you have to install manually.

> so I think it's safe to say the issue doesn't happen on Windows.

maybe Windows driver have better code to manage Wifi device to handl better on bluetooth interference. Or on that your usage on Windows, Windows have not discovering bluetooth all the time. But only when so say Windows to search for BT device.

> But what's very interesting is that once I connected a device (my mouse, an MX Master) speeds improved again! Here are the results:

so maybe it is true according from [this](https://github.com/linux-surface/linux-surface/issues/711#issuecomment-1196237460) saying that BT in discovering mode disturb WiFi. So you have to ensure that Linux have not put BT in discovering mode.

**Comment by Majiir** [comment]
No, `bluetoothctl show` reports `Discovering: no` when this issue occurs.

**Comment by jonas2515** [comment]
> Bluetooth on, no devices connected: 3.1 Mbps down / 22.3 Mbps up
> Bluetooth on, mouse connected: 93 Mbps down / 39 Mbps up

Aha, so this is the interesting part here. I bet your controller is scanning for BLE devices when disconnected because there's a BLE device paired. Try setting your controller mode to BR/EDR only in /etc/bluetooth/main.conf `ControllerMode = bredr` and restart bluetoothd.

**Comment by jonas2515** [comment]
Also, if you're on a Surface Pro 3, we're not applying any quirks for your card (see https://github.com/linux-surface/kernel/blob/v6.5-surface-devel/drivers/net/wireless/marvell/mwifiex/pcie_quirks.c) because nobody showed up to test things on SP3 back in the day. If you care about support for your device, you might want to add the entry to pcie_quirks.c and test if the same quirks are needed.

Would also be interesting if your kernel has https://github.com/linux-surface/kernel/pull/88 applied (and the output of `lsusb` to be sure that the USB ID is actually matching).

**Comment by Majiir** [comment]
Thanks for your help. Yes, I have a Surface 3 Pro.

My kernel does have that lescan interval patch applied, but the USB ID from `lsusb` is different (`204b` vs. `204c`):

```
Bus 001 Device 007: ID 1286:204b Marvell Semiconductor, Inc. Bluetooth and Wireless LAN Composite Device
```

I tried adding an entry for `204b` with `BTUSB_LOWER_LESCAN_INTERVAL` to `btusb.c` and also adding `Surface 3 Pro` to `mwifiex/pcie_quirks.c`. The kernel log shows the quirks were applied, but wifi is still much slower when bluetooth is enabled.

I also noticed that the 8897 is not present [here](https://github.com/linux-surface/kernel/blob/v6.5-surface-devel/drivers/net/wireless/marvell/mwifiex/usb.h#L16-L23) but I'm not sure what, if anything, needs to be added there.

Patch I applied:

```diff
diff --git a/drivers/bluetooth/btusb.c b/drivers/bluetooth/btusb.c
index 6c55e59790ae..4deabfa0d67e 100644
--- a/drivers/bluetooth/btusb.c
+++ b/drivers/bluetooth/btusb.c
@@ -468,7 +468,8 @@ static const struct usb_device_id blacklist_table[] = {
        { USB_DEVICE(0x1286, 0x2044), .driver_info = BTUSB_MARVELL },
        { USB_DEVICE(0x1286, 0x2046), .driver_info = BTUSB_MARVELL },
        { USB_DEVICE(0x1286, 0x204e), .driver_info = BTUSB_MARVELL },
-       { USB_DEVICE(0x1286, 0x204c), .driver_info = BTUSB_LOWER_LESCAN_INTERVAL },
+       { USB_DEVICE(0x1286, 0x204b), .driver_info = BTUSB_MARVELL | BTUSB_LOWER_LESCAN_INTERVAL },
+       { USB_DEVICE(0x1286, 0x204c), .driver_info = BTUSB_MARVELL | BTUSB_LOWER_LESCAN_INTERVAL },
 
        /* Intel Bluetooth devices */
        { USB_DEVICE(0x8087, 0x0025), .driver_info = BTUSB_INTEL_COMBINED },
diff --git a/drivers/net/wireless/marvell/mwifiex/pcie_quirks.c b/drivers/net/wireless/marvell/mwifiex/pcie_quirks.c
index 99b024ecbade..d23dbf967272 100644
--- a/drivers/net/wireless/marvell/mwifiex/pcie_quirks.c
+++ b/drivers/net/wireless/marvell/mwifiex/pcie_quirks.c
@@ -7,6 +7,16 @@
 
 /* quirk table based on DMI matching */
 static const struct dmi_system_id mwifiex_quirk_table[] = {
+       {
+               .ident = "Surface Pro 3",
+               .matches = {
+                       DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+                       DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 3"),
+               },
+               .driver_data = (void *)(QUIRK_FW_RST_D3COLD |
+                                       QUIRK_DO_FLR_ON_BRIDGE |
+                                       QUIRK_NO_BRIDGE_D3),
+       },
        {
                .ident = "Surface Pro 4",
                .matches = {
```

> Try setting your controller mode to BR/EDR only in /etc/bluetooth/main.conf ControllerMode = bredr and restart bluetoothd.

That seems to work around the issue, but it makes my mouse unusable.

Do you know how I can check the active `le_scan_interval` and `le_scan_window` values?

**Comment by jonas2515** [comment]
> My kernel does have that lescan interval patch applied, but the USB ID from lsusb is different (204b vs. 204c):

Ahh that's good to know, probably an older hardware revision of the same card with who knows what extra bugs :/

> I also noticed that the 8897 is not present [here](https://github.com/linux-surface/kernel/blob/v6.5-surface-devel/drivers/net/wireless/marvell/mwifiex/usb.h#L16-L23) but I'm not sure what, if anything, needs to be added there.

That's as it should be, yes. The wifi interface of the card is connected via PCIe on Surface devices, whereas the bluetooth interface of the same card is connected on the USB bus. This is a very common thing to do for combined wifi+bluetooth cards (bluetooth is a lot less complex and more standardized when it comes to host<->card communication, it also needs less bandwith, so it makes sense to use an easier bus for that).

The patches you applied look correct to me, you can confirm the `le_scan_interval` and `le_scan_window` by looking at the output of `sudo btmon` during startup of the bluetooth device (eg. when restarting bluetoothd). There should be a line `HCI Command: LE Set Extended Scan Parameters` with `Type: Passive` somewhere.

You can also try out different values by setting the hex values as `ScanIntervalAutoConnect` and `ScanWindowAutoConnect` in /etc/bluetooth/main.conf.

> That seems to work around the issue, but it makes my mouse unusable.

Hmm, sounds like your mouse is using bluetooth LE then, interesting. The problem with LE is that autoconnection works in  a way where the host needs to scan: Peripherals advertise themselves on the air regularly and the host scans regularly to pick up those advertisements and then connect. So after your laptop picks up the BLE advertisement of your mouse, it send s a connection request to the mouse. With BLE (unlike BR/EDR) the peripheral never connects to the host, so you need to be scanning on the advertisement channels at a somewhat regular interval as long as you're not connected (see https://punchthrough.com/manage-ble-connection/ for a good writeup on the topic).

So yeah, if you can't find passive LE scanning parameters that makes auto-connect for your mouse work while at the same time having sensible wifi speeds, it should be possible to disable passive LE scanning completely and just initiate an active scan whenever you want your mouse to connect. That should be possible either by opening bluetooth settings or telling bluetoothd to connect to the mouse (in case of LE devices that also just makes bluetoothd switch to active scanning in order to try harder to pick up any advertisements).

**Comment by Majiir** [comment]
The scan parameters are being set correctly when the device initializes:

```
< HCI Command: LE Set Scan Parameters (0x08|0x000b) plen 7
        Type: Passive (0x00)
        Interval: 250.000 msec (0x0190)
        Window: 6.250 msec (0x000a)
        Own address type: Public (0x00)
        Filter policy: Accept all advertisement (0x00)
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Scan Parameters (0x08|0x000b) ncmd 1
        Status: Success (0x00)
```

When I `bluetoothctl scan le`, something else is changing the scan parameters back:

```
< HCI Command: LE Set Scan Parameters (0x08|0x000b) plen 7
        Type: Active (0x01)                                                                                                                                                                                                                   
        Interval: 11.250 msec (0x0012)                                                                                                                                                                                                        
        Window: 11.250 msec (0x0012)                                                                                                                                                                                                          
        Own address type: Random (0x01)                                                                                                                                                                                                       
        Filter policy: Accept all advertisement (0x00)                                                                                                                                                                                        
> HCI Event: Command Complete (0x0e) plen 4
      LE Set Scan Parameters (0x08|0x000b) ncmd 1                                                                                                                                                                                             
        Status: Success (0x00)                                                                                         
```

The parameters seem to flip back and forth every 5-10 seconds. I don't understand how the various scan and discovering states work, but it looks like the `0x12` values come from [here](https://github.com/linux-surface/kernel/blob/53cb9482c87656c069fb0e542935f6cdf31f3645/net/bluetooth/hci_core.c#L2493-L2494).

The following patch successfully changes those values, and wifi performance is good during LE scanning and when no LE devices are connected:

```diff
--- a/drivers/bluetooth/btusb.c
+++ b/drivers/bluetooth/btusb.c
@@ -3966,6 +3966,8 @@ static int btusb_probe(struct usb_interface *intf,
        if (id->driver_info & BTUSB_LOWER_LESCAN_INTERVAL) {
                hdev->le_scan_interval = 0x0190;
                hdev->le_scan_window = 0x000a;
+               hdev->le_scan_int_discovery = 0x0180;
+               hdev->le_scan_window_discovery = 0x0009;
        }
 
        if (IS_ENABLED(CONFIG_BT_HCIBTUSB_MTK) &&
```

Something occasionally changes the scan interval and window to `0x60`, so I might also need to set `le_scan_int_adv_monitor`, `le_scan_int_connect`, etc. I don't yet understand these parameters well enough to submit a patch, but this seems to have solved my problem for now. Thank you!

**Comment by qzed** [comment]
@jonas2515 Would it make sense to add this to our wifi/bt patches?

**Comment by jonas2515** [comment]
I was gonna check in bluetoothd if they changed the passive scanning code or something, in theory the parameters for "discovery" should only be used when "aggressively" scanning for LE devices while the bluetooth settings are open. At other times, I think the "passive" parameters should be used. So in theory this patch should not be necessary.

**Comment by c0d3h01** [comment]
I was having this same problem on all linux distro

**Comment by tfpsly** [comment]
Old ticket, but still opened and top result in web searches, so...
On my gen1 Surface Laptop (model 1769, with a Marvell 88W8897), disabling `btcoex` on module `mwifiex_pcie` seems to help: WiFi speed stays closer to max speed when BT is disabled, drops to <5Mbps still occur sometimes but recover on their own.
```
echo "options mwifiex_pcie btcoex_enable=0" | sudo tee /etc/modprobe.d/mwifiex.conf
sudo modprobe -rf mwifiex_pcie
sudo modprobe mwifiex_pcie
```

WiFi speed before: BT off 130Mbps ; BT on no BT-LE around 66Mbps ; BT-LE buds drops down to 300Kbps for long times
WiFi speed after disabling btcoex, with BT-LE earbuds: 60% of time 130 Mbsp, 40% drops to 60Mpbs, and rare drops to 1Mbps.

All tested from the same spot in my home, minutes apart.

**Comment by tfpsly** [comment]
The Surface kernel alters btusb.c to have longer delay between LE Bluetooth scans. It helps a lot, but after a while I still have WiFi cutouts.
One can alter the BT scan settings in /etc/bluetooth/main.conf - look for "LE scanning", and change the ScanInterval* and ScanWindow*. I'm testing with the ScanInterval at the max value 0x400, and with a lower ScanWindow* of 0x003 and it seems to help a lot.
Then `systemctl bluetooth restart` to apply the changes.
The only downside is that I have to manually reconnect my Bluetooth LE earbuds; ie: when putting them in the case and back in my ears, they won't auto-reconnect. But I'll take that over a sluggish WiFi connection.
